Silver rupee shuffle

rrealmuto opened the pull request:
This PR adds a setting to shuffle silver rupees into the item/location pool. 

When enabled, silver rupees for each silver rupee puzzles in the game will be added to the item pool, and the location of vanilla silver rupee locations will be replaced with shuffled items. The number of silver rupees for each puzzles in the pool will be equal to the number of rupees normally in the puzzle (5 for all vanilla silver rupee puzzles, varies a bit for MQ). Upon collecting all of the silver rupees for a puzzle, the switch flag for that puzzle is set, enabling the player to move passed the puzzle.

Silver rupees have the same shuffle options as other dungeon options and behave as follows: 
'Remove': Silver Rupees are removed and the puzzles are solved. This will add a small amount of money and refill items to the pool.
'Vanilla': Silver Rupees will appear in their vanilla locations. You will have to collect all of a set in one go to solve a puzzle.
'Own Dungeon': Silver Rupees can only appear in their respective dungeon.
'Overworld Only': Silver Rupees can only appear outside of dungeons.
'Any Dungeon': Silver Rupees can only appear in a dungeon, but not necessarily the dungeon they are for.
'Regional': Silver Rupees can only appear in regions near the original dungeon (including the dungeon itself or other dungeons in the region).
'Anywhere': Silver Rupees can appear anywhere in the world.

The dungeon key info d-pad menu is updated to show the number of rupees that have been collected for each puzzle. The color of the numbers corresponds to a color of the location displayed in the message when the silver rupee is collected.

![silver_rupee_menu](https://user-images.githubusercontent.com/5086838/205453496-9cd12e68-b6c2-4028-af00-388c65711f7e.png)

ETR-BTF commented:
It should be pointed out that a key for the colors on the info screen can be found on the wiki already: https://wiki.ootrandomizer.com/index.php?title=Silver_Rupees

fenhl commented:
This setting is playable as-is and has been extensively tested on both Dev-M and my branch, but some nice-to-haves currently missing are:

- [ ] Silver rupee pouches, the equivalent of keyrings, are implemented but the setting has not been added due to #1572.
- [ ] Also due to #1572, silver rupee collection text boxes currently don't tell you how many of that puzzle you've already collected.
- [ ] Collecting a shuffled silver rupee currently plays the same sound effect that plays when rupees are found in a chest. Ideally it would play the appropriate sound effect depending on whether or not it's the final one that unlocks the puzzle.
- [ ] The silver rupee count for the current room, if any, should be displayed in the lower left corner of the HUD, above small keys.

ETR-BTF commented on SettingsList.py:
```diff
@@ -3455,6 +3455,61 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+            item pool.
+
+            Silver Rupees are grouped into sets of 5 (except for some
+            Master Quest dungeons, which have sets of other amounts), each
+            of which permanently unlocks something in a dungeon once all
+            the rupees in that set are collected. Hints will only tell you
+            the dungeon a Silver Rupee corresponds to, but upon collecting
+            it, you will be told the exact room.
+            The vanilla locations of Silver Rupees hold shuffled items.
+
+            'Remove': Silver Rupees are removed and the puzzles are
+            solved. This will add a small amount of money and
+            refill items to the pool.
+
+            'Vanilla': Silver Rupees will appear in their vanilla
+            locations. You will have to collect all of a set in one go to
+            to solve a puzzle.
+
+            'Own Dungeon': Silver Rupees can only appear
+            in their respective dungeon.
+
+            'Overworld Only': Silver Rupees can only appear
+            outside of dungeons.
+
+            'Any Dungeon': Silver Rupees can only appear in a
+            dungeon, but not necessarily the dungeon they are for.
+
+            'Regional': Silver Rupees can only appear in regions 
+            near the original dungeon (including the dungeon 
+            itself or other dungeons in the region).
+            <a href="https://wiki.ootrandomizer.com/index.php?title=Hints#Hint_Regions" target="_blank">The Wiki has a list of corresponding regions here.</a>
+
+            'Anywhere': Silver Rupees can appear
+            anywhere in the world.
+        ''',
+        shared         = True,
+        gui_params     = {
+            'randomize_key': 'randomize_settings',
+        },
+    ),
```
Since the SettingsList.py order is also reflected in the spoiler log, I think that this setting would fit better in between the "Key Rings" and "Shuffle Maps & Compasses" settings (lines 3033-3107).

rrealmuto replied:
Done

mracsys commented on ASM/c/dungeon_info.c:
```diff
@@ -458,62 +479,126 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
+                        if (silver_rupee_puzzles[puzzle_idx] == (uint8_t) -1) break;
+                        silver_rupee_data_t var = silver_rupee_vars[silver_rupee_puzzles[puzzle_idx]][CFG_DUNGEON_IS_MQ[d->index]];
+                        uint8_t count = extended_savectx.silver_rupee_counts[silver_rupee_puzzles[puzzle_idx]];
+                        int puzzle_left = left + icon_size + padding + font_sprite.tile_w * (2 * puzzle_idx - 1);
```
Additional padding between silver rupee counts is helpful for MQ Shadow where counts go up to 10. The suggestion is what I have on my branch. Also don't forget the A button menu with the different font size, ref. https://github.com/mracsys/OoT-Randomizer/blob/Dev-M/ASM/c/dungeon_info.c#L438
```suggestion
                        int puzzle_left = left + font_sprite.tile_w * (2 * puzzle_idx) + padding * 4 * puzzle_idx;
```

r0bd0g commented:
It'll be a while before I get around to reviewing this.

fenhl commented:
Apparently the logic for MQ Shadow invisible spikes is completely wrong.

fenhl commented:
> Apparently the logic for MQ Shadow invisible spikes is completely wrong.

Addressed in rrealmuto#63.

r0bd0g commented:
~ WIP Review ~

(Feel free to give thoughts on anything that I bring up in here, even if I haven't gone in-game to investigate yet.)

Don't forgot to add this shuffle to Hell Mode?

Nitpick: I'd appreciate a more consistent ordering between the different types of checks? I dunno it's just a bit inconsistent at times. For example between the two Shadows, one lists the chest the rupees normally open before the rupees themselves, and one lists it after. Usually I do main checks at the top, and skulltulas and drops at the bottom, just to help those more commonly important locations stand out a bit.

I'm not seeing any reference to Boomeranging silver rupees anywhere in this (one exception). I just want to ask to make sure: Is it possible to Boomerang the silver rupee checks? If so I think I would personally leave that out of logic and say that it's currently a technical limitation that it's possible to do that. I wouldn't want to see the requirements to do checks altered in such a significant way. One bit of awkwardness is maybe in Ice Cavern where there would be silver rupees and normal freestandings in the same room, and some could be Boomeranged and some couldn't, but... I guess that's fine, just deal with it I guess.

---

MQ DC:
- I suggest adding a middle region for Dodongos Cavern MQ Staircase Silver Rupee Vines. It's not strictly necessary but I think it'll be overall cleaner like this. And then you can go ahead then and backspace the added comment.

MQ Ganon's:
- To-Do: Check if all of the Water Trial MQ Rupees are safe as child.
- To-Do: Check the age and tunic logic on all Fire Trial MQ rupees.
- To-Do: Check tunic logic on reaching the final door.
- The trick to reach the end of Fire Trial with just Hookshot requires Golden Gauntlets also.

Ganon's:
- To-Do: Check age/tunic logic in Fire Trial.
- (To-Do: Confirm this was not intentional.) With Silver Rupee shuffle off you should probably hard require the tunic to give time to collect all of the rupees in one go.
- Ganons Castle Spirit Trial Crystal Switch Chest: You're going to need to check for a weapon in order to hit this switch and spawn the chest. Used to be you had to either be able to jumpslash to get into this room. I'm not sure what all will work, probably can_jumpslash or has_explosives? (Or if a bomb explosion won't reach directly, maybe it'll have to be has_bombchus instead of any explosives.) (Slingshot extension would work. Not sure if there could be a Boomerang trick. These wouldn't be worth having.)

MQ GTG:
- To-Do: Uhhhhh lava room silver rupees uhhhh I have no idea... so I'm just gonna rewrite it. And see if the result matches. I don't understand why some of the rupees check for having all of the silver rupees already. I'm not sure how coming out of the water room door would help? (And I don't remember clearing the room making the flames go away or something?).
- To-Do: Check child collecting the rupee behind the icicles just in case you can't get through. I think explosives don't work on them, can't remember for sure, so whenever icicles as child come up I usually have to at least take a look.
- Gerudo Training Ground Stalfos Room: I think you don't actually need hookshot or trick to reach this room in MQ b/c there is a path that doesn't go past any flame walls. I might suggest breaking the vanilla trick into two now that the two parts of it accomplish different things. Perhaps the Floormaster grab is not so much harder in MQ that the two Floormaster grabs can be combined into one trick? And then add a new trick for sidehopping past the flame wall in vanilla.

GTG:
- To-Do: These lava room ones aren't as crazy as MQ GTG's but I feel the same where I pretty much have to rewrite it before I can understand it.
- Gerudo Training Ground Lava Room Silver Rupee Hookshot Target: I have no idea why of all of the silver rupees, only this one is checking for Boomerang? Can child even get close enough to Boomerang it even if it were possible to Boomerang them?
- There are trick jumps to reduce some of the item requirements for getting around this lava room. I never added it before because it was obsoleted by simply turning on the lens trick anyway. I probably should pay attention to that when structuring this area, in both versions of the dungeon. (Technically in MQ you can navigate almost the entire room while the flame circles are there, but that's stupid, so we're not going to worry about that, lol. Except there was one pretty easy instance that was kind of in logic before, gated behind the Hookshot-only trick for the slopes rooms. Since you might not logically need Longshot to get through there anymore, maybe changes the equation a little bit on how to deal with that instance.)
- To-Do: I want to check if scales can reach any of these underwater silver rupees.

Ice:
- To-Do: Check child getting the rupee behind the icicles.

MQ Shadow:
- To-Do: Check if any of the invisible blades silver rupees are obtainable without getting within the radius of the blades.
- The logic for the invisible spikes room is completely wrong. It's the Stalfos chest that needs the rupees. Invisible blades chest should be free. The Hookshot check on the Stalfos chest is because several of the rupees will require Hookshot, but none of them are currently checking for it.

Shadow:
- Shadow Temple GS Single Giant Pot: I don't think Hookshot will be needed for this. It was just there previously for collecting the rupees.

MQ Spirit:
- (To-Do: Test this.) Spirit Temple MQ Lobby and Lower Adult Silver Rupee Water Jet: Here's a dumb idea that might be safe for default logic if it works. I think if the chest that this rupee normally spawns is already there? You might be able to Longshot through the water jet, collecting the rupee along the way, and thus skipping the need for Hammer. I previously tried this using the back right chest, but the geometry of the room was such that I couldn't get an angle on it. I'm not sure how easy or not it might be so see where the chest is through the graphics of the water jet.

Spirit:
- Any way you can kill the skulltula, you can get it just by climbing on the gate around it and just getting a little like that it will collect. Certainly the assumption is that players would collect the rupees and that climb on the correct side of the gate, though. Perhaps this is classic enough that we should implement a trick right away.
- But if you come in and the gate is already down, there are a few more ways to kill that skulltula than just projectiles. Din's should work no problem. I think even sticks, jumpslashing it through the gate is probably no problem. K Sword might be too short to reach without a bit of tech though? Would need to test. If K Sword is a bit tricky, might be worth combining it into the new trick.
- Nitpick: On sun block room child route, can you move the silver rupees check next to the sticks check? First line is supposed to be age and the items, and then 2nd line is room access requirement for that age.
- All of the sun block room silver rupees should have logic that matches the "free" pots in this region: "(Small_Key_Spirit_Temple, 3) or spirit_temple_shortcuts or has_explosives or ((Small_Key_Spirit_Temple, 2) and bombchus_in_logic and not entrance_shuffle)"

fenhl commented:
You may find the comments on these PRs useful for logic in the vanilla dungeons:

* [Ice](https://github.com/rrealmuto/OoT-Randomizer/pull/26)
* [Well](https://github.com/rrealmuto/OoT-Randomizer/pull/28)
* [Shadow](https://github.com/rrealmuto/OoT-Randomizer/pull/29)
* [GTG](https://github.com/rrealmuto/OoT-Randomizer/pull/30)
* [Spirit](https://github.com/rrealmuto/OoT-Randomizer/pull/32)

As for boomerang usage, that was indeed possible when I was reviewing the logic, not sure if it's still the case. I didn't realize it was possible until I was reviewing the GTG lava room, and I guess I forgot about it again later. Does the boomerang work on pot drops in vanilla? What about freestanding rupees?

r0bd0g commented:
Boomerang works on pots and freestandings. Does not normally work on Silver Rupees.
(I would very much prefer the mechanics of the checks to not change in this way.)

fenhl commented:
I didn't realize being able to boomerang freestanding rupees was a thing in vanilla. In that case I think the ideal solution would be to remove that ability from silver rupees. Not sure if that's doable though.

r0bd0g commented:
Review updated. Still to go: GTG lava rooms, learn a couple tricks, have a couple of discussions about tricks and timers and such.

r0bd0g commented:
Review completed. There are a lot of things I could use somebody please to comment on lol. If you don't comment now I'm just going to have to go around and bug people about it later.

How much of the requested changes would you like me to write?

fenhl commented:
A PR would be great, less potential for miscommunication.

r0bd0g commented:
It says I'm using the wrong version of node. So I'm not going to be writing the PR.

r0bd0g commented:
Should instances of shuffle_silver_rupees in the logic actually be shuffle_silver_rupees != "vanilla"?

I'm not sure how I feel about 'vanilla locations' as a setting b/c shuffled silver rupee behaviour at vanilla locations would be quite different logically from what the setting actually does with the normal vanilla behaviour. I think I would just call it 'off' or something.

fenhl commented:
> Should instances of shuffle_silver_rupees in the logic actually be shuffle_silver_rupees != "vanilla"?

https://github.com/TestRunnerSRL/OoT-Randomizer/blob/7638c42804927d00d963531324635edf7032360e/World.py#L52

> I'm not sure how I feel about 'vanilla locations' as a setting b/c shuffled silver rupee behaviour at vanilla locations would be quite different logically from what the setting actually does with the normal vanilla behaviour. I think I would just call it 'off' or something.

That's fair, I wouldn't be opposed to renaming the option.

r0bd0g commented:
That's all I've got for logic.
What else is left for this PR?

rrealmuto commented:
I'm still reviewing a PR from fenhl that needs to be merged before this is.

fenhl commented:
Another thing we should probably do is allow silver rupees that don't do anything based on the settings (e.g. BotW basement if dungeon pots are unshuffled or lensless BotW is on) to be foolish by adding them to [this code](https://github.com/TestRunnerSRL/OoT-Randomizer/blob/0883753b248bc3e450b529cdff4588ffbb6a096f/World.py#L1198-L1224).

rrealmuto commented:
Sorta related to what r0b was asking about the "vanilla" setting... Should logic that checks for a particular count of silver rupees be changed to first check that the silver rupee setting is enabled? Currently the playthrough requires actually requires collecting each silver rupee, which is a bit different logically than collecting all of the silver rupees. Ex.

`"Ganons Castle Fire Trial Ending": "
                is_adult and Longshot and (not shuffle_silver_rupees or (Silver_Rupee_Ganons_Castle_Fire_Trial, 5)) and
                (logic_fewer_tunic_requirements or Goron_Tunic)"`

r0bd0g commented:
My logic changes have gone missing? Don't know how any of this works lol. I noticed b/c your example logic there is using the old logic as a base, which has the same bug with requiring the tunic that I'd tried to fix.

rrealmuto commented:
Lol whoops. Fixed.

r0bd0g commented:
Need to delete all instances of and not entrance_shuffle from the Spirit file now that the bombchu changes are in.

r0bd0g commented:
I noticed while playing the hitbox for the fire trial rising platform rupee is significantly lower when shuffled. You're supposed to have to backflip to reach it. I really worry about the altered hitboxes, I dunno what else might be lurking. (I suspect vanilla spirit trial armos jump is probably a lot easier also. What else might there be that is significantly affected?)

rrealmuto commented:
So if it's a big enough issue, it may be easy enough to modify the collision to be the same as original silver rupees. Already have a hack in place that modifies the collider to make it not boomerangable so it shouldn't be too much to change the dimensions. I'll have to take a look at the differences between the two. Will let you know

r0bd0g commented:
I worry also about some of the iffier rupees to touch, where I only tested them on gz, like gold scale the toilet rupee for example. I really think the hitboxes should match even if it doesn't cause any issues, but honestly the fact that you don't have to backflip where you normally would is already big enough for me.

fenhl commented on data/World/Spirit Temple.json:
```diff
@@ -117,12 +136,18 @@
+            "Spirit Temple Sun Block Room Silver Rupee Center Front" : "True",
+            "Spirit Temple Sun Block Room Silver Rupee Center Back" : "True",
+            "Spirit Temple Sun Block Room Silver Rupee Right Front" : "True",
+            "Spirit Temple Sun Block Room Silver Rupee Right Back" : "True",
```
These 5 locations are duplicates and should be removed.

r0bd0g replied:
Why's it always gotta be the spirit file... Last time, for pots, the shield pot PR came through around the same time and caused some regressions.

r0bd0g commented:
For the whole foolish thing, what about light trial rupees?

rrealmuto commented:
I just finished up the fix for the silver rupee collisions so that they should behave identical to regular silver rupees when the player interacts with them. Also cleaned up that duplicate entry in the Spirit Temple JSON.

Anything else before marking this ready?

fenhl commented:
Seems ready to me.

r0bd0g commented:
(I have no idea how those duplicates in the Spirit logic got there. I didn't add them.)

rrealmuto commented:
> (I have no idea how those duplicates in the Spirit logic got there. I didn't add them. Bad merge somewhere along the way. Happens every time.) (There is at least one logic bug fixed from main branch in here.)

Fenhl pointed them out a while ago. Happened when I merged the logic fixes. All good 😊

fenhl commented:
Could you change the ordering of the well rupees back to what it was before? I think it makes more sense to list the bottom ones first, since you're normally going to be collecting the top ones last.

rrealmuto commented:
Does this need to be squashed/rebased before merging?

Cuphat commented:
I'm not sure why the GitHub action hasn't run unittests on this, but this PR currently fails the `test_fix_broken_drops` unittest.

```________________________________________ TestPlandomizer.test_fix_broken_drops _________________________________________
[gw4] linux -- Python 3.11.3 /home/cuphat/.pyenv/versions/3.11.3/bin/python3.11

self = <Unittest.TestPlandomizer testMethod=test_fix_broken_drops>

    def test_fix_broken_drops(self):
        # Setting off
        distribution_file, spoiler = generate_with_plandomizer("plando-fix-broken-drops-off")
        self.assertEqual(len([sphere for sphere in spoiler[':playthrough'].values() if 'Child Spirit Temple Deku Shield Pot' in sphere]), 0)

        # No deku shield available, fail to generate
        self.assertRaises(ShuffleError, lambda : generate_with_plandomizer("plando-fix-broken-drops-bad", max_attempts=1))

        # Deku shield available only via spirit shield pot
        distribution_file, spoiler = generate_with_plandomizer("plando-fix-broken-drops-good")
>       self.assertEqual(len([sphere for sphere in spoiler[':playthrough'].values() if 'Child Spirit Temple Deku Shield Pot' in sphere]), 1)
E       AssertionError: 0 != 1

Unittest.py:537: AssertionError

rrealmuto commented:
> I'm not sure why the GitHub action hasn't run unittests on this, but this PR currently fails the `test_fix_broken_drops` unittest.
> 
> ```
> [gw4] linux -- Python 3.11.3 /home/cuphat/.pyenv/versions/3.11.3/bin/python3.11
> 
> self = <Unittest.TestPlandomizer testMethod=test_fix_broken_drops>
> 
>     def test_fix_broken_drops(self):
>         # Setting off
>         distribution_file, spoiler = generate_with_plandomizer("plando-fix-broken-drops-off")
>         self.assertEqual(len([sphere for sphere in spoiler[':playthrough'].values() if 'Child Spirit Temple Deku Shield Pot' in sphere]), 0)
> 
>         # No deku shield available, fail to generate
>         self.assertRaises(ShuffleError, lambda : generate_with_plandomizer("plando-fix-broken-drops-bad", max_attempts=1))
> 
>         # Deku shield available only via spirit shield pot
>         distribution_file, spoiler = generate_with_plandomizer("plando-fix-broken-drops-good")
> >       self.assertEqual(len([sphere for sphere in spoiler[':playthrough'].values() if 'Child Spirit Temple Deku Shield Pot' in sphere]), 1)
> E       AssertionError: 0 != 1
> 
> Unittest.py:537: AssertionError
> ```

I'm pretty sure that unit test is broken. Might as well call it fix_broken_unit_test

mracsys commented:
I have a fix for this I can PR later. The Spirit Temple pot shield location was moved to another region, and the full name in the unit test includes the old region name.

fenhl commented:
I'm guessing the GitHub actions isn't running because of the merge conflicts?

cjohnson57 reviewed (approved):
Everything looks good to me, and there's been plenty of testing

cjohnson57 reviewed (changes requested):
On second thought, while this is pretty much ready to merge, I will request a couple things first:

* Rebase/resolve merge conflicts
* Support the changes in #1958 and #1968 in the new system (I was going to just do this myself when merging but got too scared of breaking things)

Once these are done I'll merge within a day or 2

rrealmuto committed:
Silver Rupee Shuffle

rrealmuto commented:
I hate you :(

rrealmuto commented:
Squashed and rebased. This probably needs more extensive testing as there were quite a few more conflicts this time around - primarily with messages for silver rupees/keys/key rings.

cjohnson57 commented:
I tested all the relevant text boxes and they were all working, however, my emulator is throwing errors at me (specifically when I land on the ground outside Link's house). Trying to figure out if it's my compiler or the branch

cjohnson57 commented:
From PJ64

![image](https://user-images.githubusercontent.com/30701749/236660781-11e5325b-c184-4ae5-bb2e-cc60710822f6.png)

![image](https://user-images.githubusercontent.com/30701749/236660789-41366a0b-6ce3-4c79-bb12-f83d082fcd46.png)

rrealmuto commented:
Is it just outside links house? Based on those error codes I probably messed something up merging things in hacks.asm.

rrealmuto committed:
Always nop in your delay slots kids

rrealmuto committed:
Add missing TCG default small key message

rrealmuto committed:
f
